homeassistant/components/binary_sensor/workday.py

```diff
@@ -0,0 +1,155 @@
+    """Setup the Workday sensor."""
+    import holidays
+
+    # Get the Sensor name from the config
+    sensor_name = config.get(CONF_NAME)
+
+    # Get the country code from the config
+    country = config.get(CONF_COUNTRY)
+
+    # Get the province from the config
+    province = config.get(CONF_PROVINCE)
+
+    # Get the list of workdays from the config
+    workdays = config.get(CONF_WORKDAYS)
+
+    # Get the list of excludes from the config
+    excludes = config.get(CONF_EXCLUDES)
+
+    # Instantiate the holidays module for the current year
+    year = datetime.datetime.now().year
+    obj_holidays = getattr(holidays, country)(years=year)
+
+    # Also apply the provience, if available for the configured country
+    if province:
+        if province not in obj_holidays.PROVINCES:
+            _LOGGER.error('There is no province/state %s in country %s',
+                          province, country)
+            return False
+        else:
+            year = datetime.datetime.now().year
+            obj_holidays = getattr(holidays, country)(prov=province,
+                                                      years=year)
+
+    # Output found public holidays via the debug channel
+    _LOGGER.debug("Found the following holidays for your configuration:")
+    for date, name in sorted(obj_holidays.items()):
+        _LOGGER.debug("%s %s", date, name)
+
+    # Add ourselves as device
+    add_devices([IsWorkdaySensor(obj_holidays, workdays,
+                                 excludes, sensor_name)], True)
+
+
+class IsWorkdaySensor(Entity):
+    """Implementation of a Workday sensor."""
+
+    def __init__(self, obj_holidays, workdays, excludes, name):
+        """Initialize the Workday sensor."""
+        self._name = name
+        self._obj_holidays = obj_holidays
+        self._workdays = workdays
+        self._excludes = excludes
+        self._state = STATE_UNKNOWN
+
+    @staticmethod
+    def day_to_string(day):
+        """Convert day index 0 - 7 to string."""
+        try:
+            return ALLOWED_DAYS[day]
+        except IndexError:
+            return None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    def is_include(self, day, now):
+        """Check if given day is in the includes list."""
+        # Check includes
+        if day in self._workdays:
+            return True
+        elif 'holiday' in self._workdays and now in self._obj_holidays:
+            return True
+
+        return False
+
+    def is_exclude(self, day, now):
+        """Check if given day is in the excludes list."""
+        # Check excludes
+        if day in self._excludes:
+            return True
+        elif 'holiday' in self._excludes and now in self._obj_holidays:
+            return True
+
+        return False
+
+    def update(self):
```

Comment:
Please change this to

```python
@asyncio.coroutine
def async_update(self):
```

Because there is no I/O being done, it doesn't make sense to run this in a thread.